### PR TITLE
v0.5.12

### PR DESCRIPTION
**Pull request by qingyun-wu**

### CODE REVIEW

File: flaml/data.py [hunk 1]
```diff
@@ -20,10 +21,13 @@ def load_openml_dataset(dataset_id, data_dir=None, random_state=0):
...
+            Can choose from ['dataframe', 'array'].
+            If 'dataframe', the returned dataset will be a Pandas DataFrame.
+            If 'array', the returned dataset will be a NumPy array or a SciPy sparse matrix.     
     Returns:
-        X_train: A dataframe of training data
-        X_test:  A dataframe of test data
+        X_train: Training data
+        X_test:  Test data
         y_train: A series of labels for training data
         y_test:  A series of labels for test data
```
Comment (sonichi) [inline flaml/data.py:32]: series or array?

File: flaml/automl.py [hunk 2]
```diff
@@ -1144,21 +1149,31 @@ def custom_metric(
...
+        else:
+            self._training_log = None
             self._search()
-            logger.info("fit succeeded")
+        logger.info("fit succeeded")
+        logger.info(f"Time taken to find the best model: {self._time_taken_best_iter}")
+        if self._time_taken_best_iter >= time_budget * 0.7 and not \
+           all(self._ever_converged_per_learner.values()):
+            logger.warn("Time taken to find the best model is larger than 0.7 of the \
+                         provided time budget and not all estimators converged.")
```
Comment (sonichi) [inline flaml/automl.py:1175]:
```suggestion
            logger.warn("Time taken to find the best model is {0:.0g}% of the "
                        "provided time budget and not all estimators' hyperparameter "
                        "search converged.".format(
                            self._time_taken_best_iter / time_budget * 100))

```

File: flaml/automl.py [hunk 3]
```diff
@@ -1144,21 +1149,31 @@ def custom_metric(
...
+            logger.warn("Consider increasing the time budget.")
```
Comment (sonichi) [inline flaml/automl.py:1176]:
```suggestion
            logger.warn("Time taken to find the best model is {0:.0g}% of the "
                        "provided time budget and not all estimators' hyperparameter "
                        "search converged. Consider increasing the time budget.".format(
                            self._time_taken_best_iter / time_budget * 100))

```

File: flaml/automl.py [hunk 4]
```diff
@@ -1365,6 +1388,15 @@ def _search(self):
                         search_state.best_loss,
                         self._best_estimator,
                         self._state.best_loss))
+                warn_threshold = 10**self._warn_count
```
Comment (sonichi) [inline flaml/automl.py:1391]: remove

File: flaml/automl.py [hunk 5]
```diff
@@ -1365,6 +1388,15 @@ def _search(self):
...
                         self._state.best_loss))
+                warn_threshold = 10**self._warn_count
+                searcher = search_state.search_alg.searcher
+                if searcher.is_ls_ever_converged and not self._ever_converged_per_learner[estimator]:
+                    self._ever_converged_per_learner[estimator] = searcher.is_ls_ever_converged
+                if all(self._ever_converged_per_learner.values()) and \
+                   self._state.time_from_start > warn_threshold * self._time_taken_best_iter:
+                    logger.warn(f"All estimator local search has converged at least once, and the total \
+                                  search time exceeds {warn_threshold} times the time taken to find the best model.")
+                    self._warn_count += 1
```
Comment (sonichi) [inline flaml/automl.py:1399]:
```suggestion
                    self._warn_threshold *= 10

```

File: flaml/automl.py [hunk 6]
```diff
@@ -1169,14 +1184,17 @@ def _search(self):
...
         self._model_history = {}
         self._config_history = {}
         self._max_iter_per_learner = 1000000  # TODO
         self._iter_per_learner = dict([(e, 0) for e in self.estimator_list])
+        self._ever_converged_per_learner = dict([(e, False) for e in self.estimator_list])
         self._fullsize_reached = False
         self._trained_estimator = None
         self._best_estimator = None
         self._retrained_config = {}
+        self._warn_count = 1
```
Comment (sonichi) [inline flaml/automl.py:1197]:
```suggestion
        self._warn_threshold = 10

```

File: flaml/automl.py [hunk 7]
```diff
@@ -1365,6 +1388,15 @@ def _search(self):
...
                         self._best_estimator,
                         self._state.best_loss))
+                warn_threshold = 10**self._warn_count
+                searcher = search_state.search_alg.searcher
+                if searcher.is_ls_ever_converged and not self._ever_converged_per_learner[estimator]:
+                    self._ever_converged_per_learner[estimator] = searcher.is_ls_ever_converged
+                if all(self._ever_converged_per_learner.values()) and \
+                   self._state.time_from_start > warn_threshold * self._time_taken_best_iter:
+                    logger.warn(f"All estimator local search has converged at least once, and the total \
+                                  search time exceeds {warn_threshold} times the time taken to find the best model.")
```
Comment (sonichi) [inline flaml/automl.py:1398]:
```suggestion
                    logger.warn("All estimator hyperparameters local search has converged at least once, "
                                 f"and the total search time exceeds {self._warn_threshold} times the time taken "
                                 "to find the best model.")

```

File: test/test_notebook_example.py [hunk 8]
```diff
@@ -1,6 +1,7 @@
-def test_automl(budget=5):
+def test_automl(budget=5, dataset_format='dataframe'):
     from flaml.data import load_openml_dataset
-    X_train, X_test, y_train, y_test = load_openml_dataset(dataset_id=1169, data_dir='test/')
+    X_train, X_test, y_train, y_test = load_openml_dataset(dataset_id=1169, data_dir='test/',
```
Comment (sonichi) [inline test/test_notebook_example.py:3]: catch OpenMLServerException?

Comment (sonichi) [reply test/test_notebook_example.py:3]: if catch here, then remove the catch in `test_automl_array()`

File: flaml/automl.py [hunk 9]
```diff
@@ -1374,6 +1391,15 @@ def _search(self):
                         search_state.best_loss,
                         self._best_estimator,
                         self._state.best_loss))
+                warn_threshold = 10**self._warn_count
+                searcher = search_state.search_alg.searcher
+                if searcher.is_ls_ever_converged and not self._ever_converged_per_learner[estimator]:
+                    self._ever_converged_per_learner[estimator] = searcher.is_ls_ever_converged
+                if all(self._ever_converged_per_learner.values()) and \
+                   self._state.time_from_start > warn_threshold * self._time_taken_best_iter:
```
Comment (sonichi) [inline flaml/automl.py:1399]:
```suggestion
                   self._state.time_from_start > self._warn_threshold * self._time_taken_best_iter:

```

File: test/test_notebook_example.py [hunk 10]
```diff
@@ -1,6 +1,12 @@
...
+def test_automl(budget=5, dataset_format='dataframe'):
     from flaml.data import load_openml_dataset
-    X_train, X_test, y_train, y_test = load_openml_dataset(dataset_id=1169, data_dir='test/')
+    from openml.exceptions import OpenMLServerException
+    try:
+        X_train, X_test, y_train, y_test = load_openml_dataset(dataset_id=1169, data_dir='test/',
+                                                               dataset_format=dataset_format)
+    except OpenMLServerException:
+        print("OpenMLServerException raised")
+
```
Comment (sonichi) [inline test/test_notebook_example.py:9]: return

File: flaml/tune/space.py [hunk 11]
```diff
@@ -76,3 +83,74 @@ def define_by_run_func(
...
+                    else:
+                        low_cost_dict = {}
+                    if choice_cost_list:
+                        choice_cost_dict = choice_cost_list[i]
+                    else:
+                        choice_cost_dict = {}
+                    add_cost_to_space(cat, low_cost_dict, choice_cost_dict)
+            if choice_cost_list:
+                if len(choice_cost_list) == len(domain.categories):
+                    domain.choice_cost = choice_cost_list
```
Comment (qingyun-wu) [inline flaml/tune/space.py:135]: does ```sample.Categorical``` have this attribute ```choice_cost ```?  Didn't find it.

File: flaml/tune/space.py [hunk 12]
```diff
@@ -76,3 +83,74 @@ def define_by_run_func(
...
+                else:
+                    domain.choice_cost = choice_cost_list[-1]
+                # sort the choices by cost
+                cost = np.array(domain.choice_cost)
+                ind = np.argsort(cost)
+                domain.categories = np.array(domain.categories)[ind].tolist()
+                domain.choice_cost = cost[ind]
+                domain.ordered = True
```
Comment (qingyun-wu) [inline flaml/tune/space.py:143]: does ```sample.Categorical``` have this attribute ```ordered ```?

File: flaml/automl.py [hunk 13]
```diff
@@ -860,6 +862,9 @@ def cat_hp_cost(self) -> dict:
             for estimator in self.estimator_list:
                 c = self._search_states[estimator].cat_hp_cost
                 configs.append(c)
+            configs.append([
```
Comment (qingyun-wu) [inline flaml/automl.py:865]: add comments in docstr about this extra element appended?

File: flaml/automl.py [hunk 14]
```diff
@@ -835,6 +834,9 @@ def low_cost_partial_config(self) -> dict:
             for estimator in self.estimator_list:
                 c = self._search_states[estimator].low_cost_partial_config
                 configs.append(c)
+            configs.append(np.argmin([
```
Comment (qingyun-wu) [inline flaml/automl.py:837]: add comments in docstr

File: flaml/tune/space.py (hunk 11 above)
Comment (sonichi) [reply flaml/tune/space.py:135]: No, I added it.

Comment (qingyun-wu) [reply flaml/tune/space.py:135]: Didn't find it. Where do you add?

Comment (sonichi) [reply flaml/tune/space.py:135]: It doesn't. I added it.

File: flaml/automl.py (hunk 13 above)
Comment (sonichi) [reply flaml/automl.py:865]: done.

File: flaml/automl.py (hunk 14 above)
Comment (sonichi) [reply flaml/automl.py:837]: done.

**Review by sonichi — APPROVED** [review]
close #156 #155 #151 
address #159

**Pull request merged**